Marktplaats categories added

Park0 opened the pull request:
Hopefully fixes #3613
Haven't tested it. I can only that the l2 category is added to the query string. It may be required to also add the l1 category.

Maybe @Pizzabroodje can test that for me :)

Park0 committed:
Update MarktplaatsBridge.php

dvikan commented:
this gonna massively increase the html on frontpage where all bridges are rendered. dont like it

Park0 committed:
Update MarktplaatsBridge.php only main categories

As the whole list is too big only main categories are used for now.

Park0 commented:
I removed the sub categories. I just copied them in. I did not see it got so big :).

Pizzabroodje commented:
> I removed the sub categories. I just copied them in. I did not see it got so big :).

Would there maybe be some way to dynamically load the L2 categories based on the L1 category that's selected? I'm guessing not as they're loaded into a constant?

Park0 commented:
> > I removed the sub categories. I just copied them in. I did not see it got so big :).
> 
> Would there maybe be some way to dynamically load the L2 categories based on the L1 category that's selected? I'm guessing not as they're loaded into a constant?

That would require some JS dependencies. If that would be acceptable than i could add that.

dvikan commented on bridges/MarktplaatsBridge.php:
```diff
@@ -57,7 +99,14 @@ class MarktplaatsBridge extends BridgeAbstract
                 'title' => 'Include the raw data behind the content',
-            ]
+            ],
+            '2' => [
```
pls find better name than `2`

dvikan commented on bridges/MarktplaatsBridge.php:
```diff
@@ -130,4 +185,64 @@ public function getName()
         return parent::getName();
     }
+
+    public static function getCategories()
```
rename

Park0 committed:
Renamed parameter 2 to sc

Renamed unused method to better reflect it usage

Park0 commented:
Done both

Park0 commented:
> > I removed the sub categories. I just copied them in. I did not see it got so big :).
> 
> Would there maybe be some way to dynamically load the L2 categories based on the L1 category that's selected? I'm guessing not as they're loaded into a constant?

@Pizzabroodje i think u better can be a separated feature request for that

Pizzabroodje commented:
> > > I removed the sub categories. I just copied them in. I did not see it got so big :).
> > 
> > 
> > Would there maybe be some way to dynamically load the L2 categories based on the L1 category that's selected? I'm guessing not as they're loaded into a constant?
> 
> @Pizzabroodje i think u better can be a separated feature request for that

Alright I'll do that.

The L1 categories can be dynamically loaded using something like this in front of the class declaration (this puts ```<optgroup label="0">``` around the options though which I don't know where that comes from). I guess this would also slow the front page down though:
```
function getCats() {
	$marktplaatsHTML = file_get_html("https://www.marktplaats.nl");
	
	$categories = array();
	
	foreach($marktplaatsHTML->find("select[id=categoryId] option") as $opt) {
		if(!str_contains($opt->innertext, "categorie")) {
			$categories += [$opt->innertext => $opt->value];
		}
	}
	
	return $categories;
}

define('CATS', getCats());
```

I don't know if Marktplaats categories ever change, but I don't think I've ever seen that happen in the 10+ years of using it. So not too important anyways. I think hardcoding the L1 categories along with a JS dependency to dynamically load the L2 categories would be best. Then it'll only do a request when selecting a category and thus wouldn't slow down the front page.

Pizzabroodje commented:
@Park0 
I wanted to use the one with hardcoded L2 categories (the first commit) myself, but get this error:
![image](https://github.com/RSS-Bridge/rss-bridge/assets/8041304/3986f9cf-a4dc-40be-affa-0a1f0ed9e56a)
Using the newer commits I don't get that.

Nevermind, fixed it. Was an issue with some square brackets not being closed properly.

Pizzabroodje commented:
> Hopefully fixes #3613 Haven't tested it. I can only that the l2 category is added to the query string. It may be required to also add the l1 category.
> 
> Maybe @Pizzabroodje can test that for me :)

Just tested this; L1 category is indeed required along with the L2 category. Only a L1 category works fine, but only a L2 category does not.

Pizzabroodje commented:
@Park0 
Also, what didn't work with the 'Fietsen en Brommers' category? For me it works fine. This is actually the category (and the sub-category 'Mountainbikes and ATB' along with it) that I used for my personal bridge a while back.


For anyone stumbling upon this thread that wants to search within a L2 category, I modified the first commit a little bit to fix some things, and added the 'Fietsen en Brommers' categories:
[MarktplaatsBridge.txt](https://github.com/RSS-Bridge/rss-bridge/files/13035013/MarktplaatsBridge.txt)

Park0 commented:
- The error about the fietsen category was for the scraping function, It may already work.
 - The categories in rss-bridge should be static in any case. The scraping of the main categories could be added to the function.
 - The error was a coding issue with the array but i think you already found that.
 - Good to know. The code now has both.
 - It was a issue with the scraping not sure what caused it. I may try it when i have some time.

Park0 committed:
Update MarktplaatsBridge.php Several fixed

Categories completed
Added a default empty one
Check if the input is not empty before using
Added helper methods to generate the categorylist

Park0 commented:
I made some small fixes. I am not aware of any issues on the current code. @dvikan if your happy with the code it can be merged.

dvikan commented:
ci fail

dvikan commented:
sorry about the ci being unclear.

it complains about public functions that should not be there.

make the helper functions non-public and it should be resolved.

Park0 committed:
Update MarktplaatsBridge.php

Set the methods to private for the CI

Park0 commented:
@dvikan Now it seems happy :)

dvikan commented:
rssbridge isnt setup nicely to handle much js stuff. also i dont want to maintain js. but we can revisit topic later
